source/testing/index.rst

```diff
@@ -0,0 +1,499 @@
+  to the database.
+* Use
+  `faster password hashing algorithms <https://docs.djangoproject.com/en/stable/topics/testing/overview/#speeding-up-the-tests>`_
+  when testing.
+* Use Continuous Integration to automatically run the whole test suite when changes are
+  made, even if developers might have skipped it.
+
+Specific Cases
+~~~~~~~~~~~~~~
+
+How to test URLs and views
+++++++++++++++++++++++++++
+
+Unfortunately, there isn’t really a good way to separate the testing
+of URL routing and the actual views, but thankfully that works out
+fine in practice. When you look at a view, you should be able to
+enumerate each type of request it's going to receive, and you can
+probably cover each of these in a separate small test.
+
+The test client is a very helpful utility Django provides to simulate
+a specific URL being hit, to send POST or GET data along with the
+request, and assert about the results usefully. The test client
+returns the Response object of the view, which even carries the
+context of any template rendering used in the view, which can be very
+helpful in testing exactly what went into rendering the response,
+without resorting to messy and error prone tests against the rendered
+HTML.
+
+Another useful tool is Django's
+`RequestFactory <https://docs.djangoproject.com/en/stable/topics/testing/advanced/#the-request-factory>`_,
+which can be used to create a Request object that can be passed
+directly to any view, bypassing the URL routing & middleware
+processes to speed things up.
+
+If bypassing middleware completely breaks things for a particular
+view, you can call specific middleware yourself on the request object first.
+
+Nevertheless, unit testing views is hard due to how many dependencies
+they introduce.
+The most valuable way to help testing views is to reduce the amount of
+code within views in the first place. Pushing functionality of
+business data logic into models and user data into forms can reduce
+most views to little more than glue connecting simpler components you
+can separately test much easier.
+
+References:
+
+* `Django: Writing and running tests <https://docs.djangoproject.com/en/stable/topics/testing/overview/#module-django.test.client>`_
+
+How to test database behaviors
+++++++++++++++++++++++++++++++
+
+Your application’s ``models.py`` should contain all of its database
+interaction, and its associated test suite will need to verify these
+are working properly. You need to test that query helpers, like
+manager methods, are giving you the right results given expected
+database contents, and that models validate, save, and are properly
+updated by other helpers. All of these require tests that actually
+work on a real database, because some behaviors can only be replicated
+accurately with a full round trip of the SQL.
+
+The Django test runner provides facilities that create a new database
+just for the testsand run migrations to create all the tables your
```

Comment:
s/testsand/tests and

Also the sentence seems a little long... maybe it could be split in 2?